src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -209,9 +210,30 @@ protected virtual void DeleteFile(string publishFilePath)
 
         protected virtual void SerializeToFile(ProjectSnapshot projectSnapshot, string publishFilePath)
         {
-            var fileInfo = new FileInfo(publishFilePath);
-            using var writer = fileInfo.CreateText();
-            _serializer.Serialize(writer, projectSnapshot);
+            // We need to avoid having an incomplete file at any point, but our project.razor.json is large enough that it will be written
+            // as multiple operations
+            var tempFilePath = string.Concat(publishFilePath, TempFileExt);
+            var tempFileInfo = new FileInfo(tempFilePath);
+
+            if (tempFileInfo.Exists)
+            {
+                tempFileInfo.Delete();
+            }
+
+            // This needs to be in explicit brackets because the operation needs to be completed
+            // by the time we move the tempfile into its place
+            using (var writer = tempFileInfo.CreateText())
+            {
+                _serializer.Serialize(writer, projectSnapshot);
+
+                var fileInfo = new FileInfo(publishFilePath);
+                if (fileInfo.Exists)
+                {
+                    fileInfo.Delete();
+                }
+            }
+
+            tempFileInfo.MoveTo(publishFilePath);
```

Comment:
I think what you want here is [Replace](https://docs.microsoft.com/en-us/dotnet/api/system.io.fileinfo.replace?view=netcore-3.1). It replaces the contents of the original file and doesn't ever delete it. This way we'll see this as file change events instead of delete/added events.

Also, the way you have it, at the end of the method the .temp file will be left standing. I suggest deleting the .temp at the end of this method and adding a Debug.Assert at the top to make sure it doesn't exist before we run serialize.

Reply:
Disclaimer about Replace: I'm not sure how it works internally. If a change event fires when it is partially written, we can't use it. But it is definitely worth testing.

Reply:
Replace would be a perfect candidate for this, but it seems the internals are going to work against us. It seems that it can perform multiple write actions on the non-temp file, and under the covers does end up firing both an Added and Removed event. On a less important note, it claims `Pass null to the destBackupFileName parameter if you do not want to create a backup of the file being replaced.`, but this is a lie. If you pass null to `destBacukupFileName` it will throw an exception with `Unable to find the specified file.:null`

Reply:
Also, why do you think the temp file will remain? That's why I used `MoveTo` instead of `CopyTo`, and when I run this locally there is no temp file unless we're in the middle of Serialization.

Reply:
> Replace would be a perfect candidate for this, but it seems the internals are going to work against us.

Ah if it also writes in batches it's useless. Ignore.

> Also, why do you think the temp file will remain?

I was thrown off by the temp file deletion at the top of the method. Given that the temp file is not supposed to exists when the method is called, do we need to `Exists() -> Delete()` condition? I'd imagine a Debug.Assert would suffice.

Reply:
@NTaylorMullen requested resiliency here, if execution was interrupted here for any reason (process crash, exception thrown) then we would be in a dirty state.

Reply:
Got it. I'd still suggest adding a `Debug.Fail` if the file exists in addition to the Delete(). We should still know about and investigate issues if we end up having inconsistent states. Also, add a comment there to indicate we don't expect the file to exist but just being defensive.

Reply:
Oh I see, you're suggesting BOTH. Both is good.